src/main/java/org/jabref/logic/openoffice/action/EditMerge.java

```diff
@@ -0,0 +1,344 @@
+        XTextRange prevRange;
+
+        ScanState() {
+            reset();
+        }
+
+        void reset() {
+            currentGroup = new ArrayList<>();
+            currentGroupCursor = null;
+            cursorBetween = null;
+            prev = null;
+            prevRange = null;
+        }
+    }
+
+    /**
+     * Decide if group could be added to state.currentGroup
+     *
+     * @param group The CitationGroup to test
+     * @param currentRange The XTextRange corresponding to group.
+     *
+     * @return false if cannot add, true if can.  If returned true, then state.cursorBetween and
+     *  state.currentGroupCursor are expanded to end at the start of currentRange.
+     */
+    private static boolean checkAddToGroup(ScanState state, CitationGroup group, XTextRange currentRange) {
+
+        if (state.currentGroup.isEmpty()) {
+            return false;
+        }
+
+        Objects.requireNonNull(state.currentGroupCursor);
+        Objects.requireNonNull(state.cursorBetween);
+        Objects.requireNonNull(state.prev);
+        Objects.requireNonNull(state.prevRange);
+
+        // Only combine (Author 2000) type citations
+        if (group.citationType != CitationType.AUTHORYEAR_PAR) {
+            return false;
+        }
+
+        if (state.prev != null) {
+
+            // Even if we combine AUTHORYEAR_INTEXT citations, we would not mix them with AUTHORYEAR_PAR
+            if (group.citationType != state.prev.citationType) {
+                return false;
+            }
+
+            if (!UnoTextRange.comparables(state.prevRange, currentRange)) {
+                return false;
+            }
+
+            // Sanity check: the current range should start later than the previous.
+            int textOrder = UnoTextRange.compareStarts(state.prevRange, currentRange);
+            if (textOrder != (-1)) {
+                String msg =
+                    String.format("MergeCitationGroups:"
+                                  + " \"%s\" supposed to be followed by \"%s\","
+                                  + " but %s",
+                                  state.prevRange.getString(),
+                                  currentRange.getString(),
+                                  ((textOrder == 0)
+                                   ? "they start at the same position"
+                                   : "the start of the latter precedes the start of the first"));
+                LOGGER.warn(msg);
+                return false;
+            }
+        }
+
+        if (state.cursorBetween == null) {
+            return false;
+        }
+
+        Objects.requireNonNull(state.cursorBetween);
+        Objects.requireNonNull(state.currentGroupCursor);
+
+        // assume: currentGroupCursor.getEnd() == cursorBetween.getEnd()
+        if (UnoTextRange.compareEnds(state.cursorBetween, state.currentGroupCursor) != 0) {
+            String msg = ("MergeCitationGroups: cursorBetween.end != currentGroupCursor.end");
```

Comment:
Can be inlined.
Please avoid technical details in error messages, thats up for debug or logger. End user must be able to understand why this fails.

Reply:
Changed to
```
                LOGGER.warn("MergeCitationGroups: cursorBetween.end != currentGroupCursor.end (during expand)");
                throw new IllegalStateException("MergeCitationGroups failed");
```

> End user must be able to understand why this fails.

It would be nice, if I did. Failures around here happened due to
  using visual order with two-column layout and/or viewing two pages side-by-side.
  This could provide an order of groups that did not respect `XText` boundaries and/or the textual order within.
  Now this should not happen, `scan` calls `frontend.getCitationGroupsSortedWithinPartitions(doc, false)` to
  get the groups.
- Turning on "Record changes" with or without "Show changes" can result in behaviour I do not really understand.
  `OOBibBase2.checkIfOpenOfficeIsRecordingChanges`  is intended to check against this.
- Something else? There might be, I am not sure. I hope not, but testing anyway, so we get at least a warning
  if the assumptions fail here. 

In the end, I cannot tell why did the test fail, only that it did.